[java] MissingOverrideRule exception when analyzing PMD under Java 9

I'm only seeing this, and sporadically, when analyzing PMD with PMD 6.3.0 under Java 9 with a full classpath

```
Exception in thread "main" java.lang.LinkageError: loader constraint violation in interface itable initialization: when resolving method "net.sourceforge.pmd.lang.xml.ast.AbstractDomNodeProxy.setUserData(Ljava/lang/String;Ljava/lang/Object;Lorg/w3c/dom/UserDataHandler;)Ljava/lang/Object;" the class loader (instance of net/sourceforge/pmd/util/ClasspathClassLoader) of the current class, net/sourceforge/pmd/lang/xml/ast/AbstractDomNodeProxy, and the class loader (instance of <bootloader>) for interface org/w3c/dom/Node have different Class objects for the type org/w3c/dom/UserDataHandler used in the signature
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3139)
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2266)
	at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.overriddenMethods(MissingOverrideRule.java:140)
	at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.getMethodLookup(MissingOverrideRule.java:112)
	at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.visit(MissingOverrideRule.java:52)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept(ASTClassOrInterfaceDeclaration.java:55)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:232)
	at net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration.jjtAccept(ASTTypeDeclaration.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:212)
	at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.visit(MissingOverrideRule.java:46)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visitAll(AbstractJavaRule.java:39)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply(AbstractJavaRule.java:27)
	at net.sourceforge.pmd.RuleSet.apply(RuleSet.java:501)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:143)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:183)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:50)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
	at net.sourceforge.pmd.processor.MonoThreadProcessor.runAnalysis(MonoThreadProcessor.java:29)
	at net.sourceforge.pmd.processor.AbstractPMDProcessor.processFiles(AbstractPMDProcessor.java:110)
	at net.sourceforge.pmd.PMD.processFiles(PMD.java:326)
	at net.sourceforge.pmd.PMD.doPMD(PMD.java:237)
	at net.sourceforge.pmd.PMD.run(PMD.java:466)
	at net.sourceforge.pmd.cli.PMDCommandLineInterface.run(PMDCommandLineInterface.java:171)
	at net.sourceforge.pmd.PMD.main(PMD.java:436)
```

@oowekyala any ideas?

New occurrence: If you execute only the pmd-xml module (e.g. `./mvnw clean install -rf :pmd-xml`), maven-pmd-plugin 3.10.0 now fails.... Note: I've now enabled the MissingOverride rule...

Stacktrace:
```
Caused by: java.lang.LinkageError: loader constraint violation: loader (instance of <bootloader>) previously initiated loading for a different type with name "org/w3c/dom/UserDataHandler"
    at java.lang.Class.getDeclaredMethods0 (Native Method)
    at java.lang.Class.privateGetDeclaredMethods (Class.java:3119)
    at java.lang.Class.getDeclaredMethods (Class.java:2268)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.overriddenMethodsRec (MissingOverrideRule.java:154)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.overriddenMethodsRec (MissingOverrideRule.java:190)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.overriddenMethods (MissingOverrideRule.java:140)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.getMethodLookup (MissingOverrideRule.java:112)
    at net.sourceforge.pmd.lang.java.rule.bestpractices.MissingOverrideRule.visit (MissingOverrideRule.java:52)
    at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept (ASTClassOrInterfaceDeclaration.java:55)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:56)
...
```

Ok, no, it doesn't matter, whether you build everything or just pmd-xml....

--- a/.gitignore
+++ b/.gitignore
@@ -10,3 +10,4 @@ bin/
 .idea
 *.patch
 */src/site/site.xml
+pmd-core/dependency-reduced-pom.xml

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -43,6 +43,7 @@ This is a minor release.
 *   ecmascript
     *   [#861](https://github.com/pmd/pmd/issues/861): \[ecmascript] InnaccurateNumericLiteral false positive with hex literals
 *   java
+    *   [#1074](https://github.com/pmd/pmd/issues/1074): \[java] MissingOverrideRule exception when analyzing PMD under Java 9
     *   [#1174](https://github.com/pmd/pmd/issues/1174): \[java] CommentUtil.multiLinesIn() could lead to StringIndexOutOfBoundsException
 *   java-bestpractices
     *   [#651](https://github.com/pmd/pmd/issues/651): \[java] SwitchStmtsShouldHaveDefault should be aware of enum types

--- a/pmd-core/pom.xml
+++ b/pmd-core/pom.xml
@@ -108,6 +108,15 @@
         <dependency>
             <groupId>jaxen</groupId>
             <artifactId>jaxen</artifactId>
+            <!--
+                Jaxen is optionally, so that it doesn't get inherited by default.
+                Jaxen is shaded into pmd-core (without org.w3c.dom.UserDataHandler).
+                Adding Jaxen again as a dependency would reintroduce the duplicated
+                org.w3c.dom.UserDataHandler interface.
+                See https://github.com/pmd/pmd/issues/1074
+                and https://github.com/pmd/pmd/pull/1201
+            -->
+            <optional>true</optional>
         </dependency>
         <dependency>
             <groupId>net.java.dev.javacc</groupId>

--- a/pmd-core/pom.xml
+++ b/pmd-core/pom.xml
@@ -60,6 +60,34 @@
                     <suppressionsLocation>pmd-core-checkstyle-suppressions.xml</suppressionsLocation>
                 </configuration>
             </plugin>
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <configuration>
+                    <artifactSet>
+                        <includes>jaxen:jaxen</includes>
+                    </artifactSet>
+                    <filters>
+                        <filter>
+                            <artifact>jaxen:jaxen</artifact>
+                            <includes>
+                                <include>org/jaxen/**</include>
+                            </includes>
+                            <excludes>
+                                <exclude>org/w3c/dom/**</exclude>
+                            </excludes>
+                        </filter>
+                    </filters>
+                </configuration>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
         </plugins>
     </build>
     <dependencies>

--- a/pmd-java/pom.xml
+++ b/pmd-java/pom.xml
@@ -108,10 +108,6 @@
     </build>
     <dependencies>
         <dependency>
-            <groupId>jaxen</groupId>
-            <artifactId>jaxen</artifactId>
-        </dependency>
-        <dependency>
             <groupId>net.java.dev.javacc</groupId>
             <artifactId>javacc</artifactId>
         </dependency>

--- a/pmd-java8/pom.xml
+++ b/pmd-java8/pom.xml
@@ -55,10 +55,6 @@
         </dependency>
 
         <dependency>
-            <groupId>jaxen</groupId>
-            <artifactId>jaxen</artifactId>
-        </dependency>
-        <dependency>
             <groupId>net.sourceforge.pmd</groupId>
             <artifactId>pmd-core</artifactId>
         </dependency>

--- a/pmd-plsql/pom.xml
+++ b/pmd-plsql/pom.xml
@@ -112,10 +112,6 @@
             <groupId>net.sourceforge.saxon</groupId>
             <artifactId>saxon</artifactId>
         </dependency>
-        <dependency>
-            <groupId>jaxen</groupId>
-            <artifactId>jaxen</artifactId>
-        </dependency>
 
         <dependency>
             <groupId>junit</groupId>

--- a/pmd-xml/pom.xml
+++ b/pmd-xml/pom.xml
@@ -66,10 +66,6 @@
     </build>
     <dependencies>
         <dependency>
-            <groupId>jaxen</groupId>
-            <artifactId>jaxen</artifactId>
-        </dependency>
-        <dependency>
             <groupId>net.sourceforge.pmd</groupId>
             <artifactId>pmd-core</artifactId>
         </dependency>

